modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C

```diff
@@ -0,0 +1,1310 @@
+               "basis (each component must be provided with a single constraint");
+
+  // check that each _constrained_species name appears in the basis
+  for (const auto & name : _mgd.basis_species_name)
+    if (std::find(_constrained_species.begin(), _constrained_species.end(), name) ==
+        _constrained_species.end())
+      mooseError("The basis species ", name, " must appear in the constrained species list");
+
+  // order the constraints in the same way as the basis species.  This makes the remainder of the
+  // code much cleaner.
+  std::vector<std::string> c_s(_num_basis);
+  std::vector<Real> c_v(_num_basis);
+  std::vector<ConstraintMeaningEnum> c_m(_num_basis);
+  for (unsigned i = 0; i < _num_basis; ++i)
+  {
+    const unsigned basis_ind = _mgd.basis_species_index.at(_constrained_species[i]);
+    c_s[basis_ind] = _constrained_species[i];
+    c_v[basis_ind] = _constraint_value[i];
+    c_m[basis_ind] = _constraint_meaning[i];
+  }
+  _constrained_species = c_s;
+  _constraint_value = c_v;
+  _original_constraint_value = c_v;
+  _constraint_meaning = c_m;
+
+  // run through the constraints, checking physical and chemical consistency
+  for (unsigned i = 0; i < _constrained_species.size(); ++i)
+  {
+    const std::string name = _constrained_species[i];
+
+    // if the mass of solvent water is provided, check it is positive
+    if (_constraint_meaning[i] == ConstraintMeaningEnum::KG_SOLVENT_WATER)
+      if (_constraint_value[i] <= 0.0)
+        mooseError("Specified mass of solvent water must be positive: you entered ",
+                   _constraint_value[i]);
+
+    // if activity is provided, check it is positive
+    if (_constraint_meaning[i] == ConstraintMeaningEnum::ACTIVITY)
+      if (_constraint_value[i] <= 0.0)
+        mooseError("Specified activity values must be positive: you entered ",
+                   _constraint_value[i]);
+
+    // if fugacity is provided, check it is positive
+    if (_constraint_meaning[i] == ConstraintMeaningEnum::FUGACITY)
+      if (_constraint_value[i] <= 0.0)
+        mooseError("Specified fugacity values must be positive: you entered ",
+                   _constraint_value[i]);
+
+    // if free molality is provided, check it is positive
+    if (_constraint_meaning[i] == ConstraintMeaningEnum::FREE_MOLALITY)
+      if (_constraint_value[i] <= 0.0)
+        mooseError("Specified free molality values must be positive: you entered ",
+                   _constraint_value[i]);
+
+    // if free moles of a mineral is provided, check it is positive
+    if (_constraint_meaning[i] == ConstraintMeaningEnum::FREE_MOLES_MINERAL_SPECIES)
+      if (_constraint_value[i] <= 0.0)
+        mooseError("Specified free mole number of mineral species must be positive: you entered ",
+                   _constraint_value[i]);
+
+    // check that water is provided with correct meaning
+    if (name == "H2O")
```

Comment:
Is this function called just at the start? If not, and it is called every iteration, we probably don't want to use a string comparison here as it is slow

Reply:
Yea, it's just called in the constructor, so string comparison shouldn't be too expensive